drivers: adc: TI CC13xx/CC26xx ADC support

niflostancu opened the pull request:
Hello, I implemented a new ADC driver for the TI CC13xx/CC26xx family.

Tested using the adc sample on my CC1352P1 LaunchXL board.
I configured the demo DTS overlay to measure the chip's VDDS voltage.

Fixes #49228

<details>
<summary>See sample logs here:</summary>

```
*** Booting Zephyr OS build zephyr-v3.1.0-3811-g3217d9737530  ***
ADC reading:
- adc@400cb008, channel 5: [03123 = 3278 mV
0:00:00.005,859] <dbg> adc_cc13xx_cc26xx: adc_cc13xx_cc26xx_channel_setup: Setup 5 acq time 9
[00:00:00.009,582] <dbg> adc_cc13xx_cc26xx: adc_cc13xx_cc26xx_isr: ISR flags 0x00000480 fifo 0x00000000
[00:00:00.009,613] <dbg> adc_cc13xx_cc26xx: adc_cc13xx_cc26xx_isr: ADC val 3123
[00:00:00.009,643] <dbg> adc_cc13xx_cc26xx: adc_cc13xx_cc26xx_isr: ISR flags 0x00000400 fifo 0x00000001
ADC reading:
- adc@400cb008, channel 5: 3124 = 3279 mV
```
</details>

Note: requires [`hal_ti` changes (PR here)](https://github.com/zephyrproject-rtos/hal_ti/pull/39).

niflostancu committed:
drivers: ti: cc13xx/cc26xx: implement ADC driver

New ADC driver for the TI CC13xx/CC26xx family.

ADC channel configurations are translated from Zephyr constants to
simplelink driverlib ones (e.g., sample times use a lookup table).

Async mode was also implemented & tested.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
boards: cc13x2/cc26x2*: add ADC sample + update docs

Add ADC driver sample overlays + update features in doc + metadata for TI
CC13x2/CC26x2 based boards (cc1352r_sensortag, cc1352r1_launchxl,
cc26x2r1_launchxl).

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

galak commented on dts/arm/ti/cc13x2_cc26x2.dtsi:
```diff
@@ -149,6 +155,14 @@
+		adc0: adc@400cb008 {
+			compatible = "ti,cc13xx-cc26xx-adc";
+			reg = <0x400CB008 0x1>;
+			interrupts = <EXP_AUX_ADC 0>;
```
just use the number, no point in having the #defines, above.   Add a comment like rtc has.

cfriedt reviewed (approved):
LGTM

cfriedt commented on west.yml:
```diff
@@ -135,7 +135,7 @@ manifest:
         - hal
     - name: hal_ti
-      revision: 29af528900d03af02a17b4b7f4fc93f19a81da2a
+      revision: pull/39/head
```
Are we still waiting for a PR in Hal?

niflostancu replied:
Well, it's a new (small) one enabling the ADC module in CMakeLists :D 
https://github.com/zephyrproject-rtos/hal_ti/pull/39

cfriedt commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,288 @@
+};
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+DEVICE_DT_INST_DEFINE(0,
```
Ah - I just noticed this here too. Please use `DT_INST_FOREACH_*()` as well

niflostancu commented:
I've switched the driver to `DT_INST_FOREACH_STATUS_OKAY`
Also added a separate commit updating the sample overlays + docs + yaml metadata for all supported boards.
Oh, and I updated `west.yml`'s `hal_ti` entry's commit ID, I hope this is the way to do it.

galak commented on dts/arm/ti/cc13x2_cc26x2.dtsi:
```diff
@@ -149,6 +149,14 @@
+
+		adc0: adc@400cb008 {
+			compatible = "ti,cc13xx-cc26xx-adc";
+			reg = <0x400CB008 0x1>;
```
nit: if you make any updates, change the case to lower `0x400cb008`

DhruvaG2000 reviewed (dismissed):
I haven't been able to test on hardware, but code LGTM.

cfriedt commented:
@niflostancu - can you please rebase? Should be good to go in

anangl commented on drivers/adc/Kconfig.cc13xx_cc26xx:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config ADC_CC13XX_CC26XX
+	bool "CC32XX ADC driver"
```
```suggestion
	bool "CC13XX/CC26XX ADC driver"
```

anangl commented on dts/bindings/adc/ti,cc13xx-cc26xx-adc.yaml:
```diff
@@ -0,0 +1,21 @@
+
+compatible: "ti,cc13xx-cc26xx-adc"
+
+include: [adc-controller.yaml, pinctrl-device.yaml]
```
Do you really need `pinctrl-device.yaml` here?

anangl commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,310 @@
+	{ 6, AUXADC_SAMPLE_TIME_5P3_US },
+	{ 11, AUXADC_SAMPLE_TIME_10P6_US },
+	{ 22, AUXADC_SAMPLE_TIME_21P3_US },
+	{ 44, AUXADC_SAMPLE_TIME_42P6_US },
```
```suggestion
	{ 43, AUXADC_SAMPLE_TIME_42P6_US },
```

anangl commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,310 @@
+/**
+ * Channels are based on ADC_COMPB_IN_* hal_ti definitions, max. index is 16 (included).
+ */
+#define CHAN_COUNT 0x10
```
This is a bit misleading, as in fact the channel count is 17. Perhaps something like `MAX_CHAN_ID` would be more adequate.

anangl commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,310 @@
+{
+	struct adc_cc13xx_cc26xx_data *data = dev->data;
+	/* get the statuses of ADC_DONE and ADC_IRQ events in order to clear them both */
+	uint32_t intStatus = (
```
Please don't use camel-case in Zephyr tree source files.

anangl commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,310 @@
+	.ref_internal = 4300, /* fixed reference: 4.3V */
+};
+
+#define CC13XX_CC26XX_ADC_IRQ_CONNECT(index)					 \
+	IRQ_CONNECT(DT_INST_IRQN(index), DT_INST_IRQ(index, priority),		 \
+		adc_cc13xx_cc26xx_isr, DEVICE_DT_INST_GET(index), 0);		 \
+	irq_enable(DT_INST_IRQN(index));					 \
```
```suggestion
	irq_enable(DT_INST_IRQN(index))
```

anangl replied:
Do you actually need a macro for this? Couldn't that content be placed directly in the `adc_cc13xx_cc26xx_cfg_func_##index` function body?

anangl commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,310 @@
+	if (sample_time_us) {
+		/* choose the nearest sample time configuration */
+		data->sample_time = adc_cc13xx_sample_times[0].reg_value;
+		for (i = 0; i < ARRAY_SIZE(adc_cc13xx_sample_times); i++) {
+			if (adc_cc13xx_sample_times[i].time_us > sample_time_us) {
+				break;
+			}
+			data->sample_time = adc_cc13xx_sample_times[i].reg_value;
+		}
```
Shouldn't an error be returned when acquisition time > 10.9 ms is requested?

niflostancu replied:
I added a check if the iteration reached the for's end to return an error. Hope that's okay.

anangl commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,310 @@
+	{ 22, AUXADC_SAMPLE_TIME_21P3_US },
+	{ 44, AUXADC_SAMPLE_TIME_42P6_US },
+	{ 86, AUXADC_SAMPLE_TIME_85P3_US },
+	{ 170, AUXADC_SAMPLE_TIME_170_US },
```
Since you are using `(adc_cc13xx_sample_times[i].time_us > sample_time_us)` in the loop below, this should be 171 (and similarly for the next 6 entries). But it actually seems that using `>=` in the condition and floor rounding of the values would be clearer.

niflostancu commented:
Thanks. I will update my PR tomorrow ;)

niflostancu commented:
Done: rebased, applied all of @anangl's suggestions, re-tested on my launchxls, nothing broken ;)

cfriedt commented:
Getting `#error unsupported board` - is something else missing in CMake or possibly in the HAL?

niflostancu committed:
tests: ADC API: add support for cc1352r* boards

Add ADC API tests for the TI CC1352R1 LaunchXL, CC2652R1 LaunchXL and
TI CC1352R SensorTag boards.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu commented:
Well, it seems that the tests in `adc_api` have hardcoded `if defined(CONFIG_BOARD_*)` conditions for all supported boards :D 
Added CC1352R* branch in there (new commit), tests should pass now ;)

cfriedt commented:
> Well, it seems that the tests in `adc_api` have hardcoded `if defined(CONFIG_BOARD_*)` conditions for all supported boards :D Added CC1352R* branch in there (new commit), tests should pass now ;)

Almost - I would suggest running tests with `twister` before pushing
```
twister -b -p cc1352r1_launchxl -p cc1352r_sensortag -p cc26x2r1_launchxl -p cc3220sf_launchxl -p cc3235sf_launchxl -T tests/drivers/adc
```

niflostancu commented:
Yep, I've just discovered `--device-testing` and I've been trying to pass the ADC tests on my CC1352* LaunchXL (instead of the sample, which worked).

I get this:
```
Running TESTSUITE adc_basic
===================================================================
START - test_adc_asynchronous_call
Samples read: 0x0c35 0xffff8000
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/adc/adc_api/src/test_adc.c:467: check_samples: (INVALID_ADC_VALUE equal to sample_value)
[1] should be filled
FAIL - test_adc_asynchronous_call in 0.020 seconds
```

It seems it's only sampling the first time, it probably gets interrupted, I need to further investigate the driver why it happens.

Unfortunately, I'll have to postpone this until next week since I'll be going AFK for several days.

cfriedt commented:
> Unfortunately, I'll have to postpone this until next week since I'll be going AFK for several days.

Good luck - looking forward to seeing this land.

cfriedt commented:
@anangl - PTAL

anangl commented on tests/drivers/adc/adc_api/boards/cc1352r1_launchxl.overlay:
```diff
@@ -0,0 +1,29 @@
+/ {
+	zephyr,user {
+		/* 0x5 is ADC_COMPB_IN_VDDS (power supply measurement) */
+		io-channels = <&adc0 0x5>;
```
The ADC API test does not use channel definitions from devicetree, so adding these overlay files is misleading.

niflostancu replied:
I didn't enable the `adc` node inside the boards `dts`, now I've done that and eliminated the overlays from the test. Hope it's okay now.

niflostancu commented on drivers/adc/adc_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,311 @@
+	{ 11, AUXADC_SAMPLE_TIME_10P6_US },
+	{ 22, AUXADC_SAMPLE_TIME_21P3_US },
+	{ 43, AUXADC_SAMPLE_TIME_42P6_US },
+	{ 86, AUXADC_SAMPLE_TIME_85P3_US },
```
Oops, done ;)
